### ISSUE TITLE
Definition of Priorities

### ISSUE THREAD

**Issue by boldtrn** [mapsforge/vtm#307]
I was playing around with the MapBox tiles and saw that major_roads have priority="1". Roads have priority of 2. Villages have priority of 3. Suburbs priority of 4. Shouldn't these be at least more important than regular roads?

**Comment by devemux86** [comment]
What do you mean?
Roads are ways lower in theme rules order, while villages / suburbs are nodes on top.

**Comment by boldtrn** [comment]
Sorry, let me rephrase.

I was talking about the text a road or a node. Here is an example:
```
            <m v="suburb" zoom-max="14">
                <caption style="bold_italic" fill="#404040" k="name" priority="4" size="17"
                    stroke="#ffffff" stroke-width="2.0" />
            </m>
```

The suburb caption has priority 4

```
    <style-text fill="#101010" id="road" k="name" priority="2" size="16" stroke="#eeeeee"
        stroke-width="2.0" />
```

Road has priority 2

So if I understand the concept correctly, if there is a conflict between different texts, the lowest priority wins. So a road label would overwrite a suburb label. This is in my opinion not what most people would expect. Maybe a major road is more important than a suburb but not a regular road. Don't you think?

I found a good example here: -33.839909,151.206551 at about zoom-level 13 - 15 (I don't remember exactly) North Sydney was overwritten by some random street. If you change the priorities North Sydney shows up as it's meant to be.

**Comment by devemux86** [comment]
Ah now I understand, have to check the code too..

Expectations could be that rule groups like city names have their own internal priorities range (0-...) being on top of other elements per theme order.
If those merge / conflict with other below rule groups like roads, then will have to see what happens.
Should we fix it somehow, change all priorities to be unique, ... ?

**Comment by boldtrn** [comment]
I found also some interesting issues with priorities and regions. City labels can overwrite region labels. Both are prio 1, but cities are defined first in the file, so probably the first match with same priorities is matched?

Maybe it would make sense to use the `sort_rank` of Mapzen as a default and only provide priorities if we want non-default behavior.

[More information on sort_rank.](https://mapzen.com/documentation/vector-tiles/layers/)

**Comment by devemux86** [comment]
I would expect that the Mapzen theme needs further adjustments.

Regarding Mapzen's [feature ordering](https://mapzen.com/documentation/vector-tiles/layers/#feature-ordering) it's one of their custom tags.
Element ordering in VTM themes (like in Mapsforge) is based on rules order and use of the `priority` tag.
Wondering how this could be tied with a data source's custom tagging like here?

**Comment by boldtrn** [comment]
> Wondering how this could be tied with a data source's custom tagging like here?

The biggest problem right now is that their ordering is inverse to priority, the higher the number with Mapzen order, the lower the priority. But in theory we might be able to use something like:

`defaultPriority=1000-sort_rank`

The defaultPriority can be overwritten by a specified priority in the mapzen.xml.

**Comment by devemux86** [comment]
Hmm Mapsforge and VTM use different priority ordering. :slightly_frowning_face: 
In [Mapsforge](https://github.com/mapsforge/mapsforge/blob/master/docs/Rendertheme.md#priorities) higher priorities are drawn first (priority 1 wins priority 0).
In VTM lower priorities are drawn first (priority 0 wins priority 1).

Regarding Mapzen's [sort_rank](https://mapzen.com/documentation/vector-tiles/layers/#feature-ordering) isn't that about only z-order, i.e. which features should appear under or over others?
Also they mention that only some of their layers have that tag.

**Comment by boldtrn** [comment]
> Regarding Mapzen's sort_rank isn't that about only z-order, i.e. which features should appear under or over others?

You are right. I thought it would be ok to assert that if a feature is drawn over another that also that label should be drawn, but this might be wrong.
